File: test/level0/schema.test.js
// The schema reader and the note checker: a note against the schema its kind names.
// Every keyword a schema holds takes a case that feeds a bad note and reads the
// finding it answers, because a case asserting nothing passes whatever the
// checker does.

import assert from "node:assert/strict";
import { test } from "node:test";
import { checkNote, readYaml } from "../../.claude/skills/level0/lib/schema.js";
import * as yaml from "../../.claude/skills/level0/lib/schema-yaml.js";
import { itemsIn } from "../../.claude/skills/level0/lib/schema-body.js";
import { SEVERITY } from "../../.claude/skills/level0/lib/schema-fault.js";
import { mintNote } from "../../.claude/skills/level0/lib/schema-mint.js";
import { kindOf, readNote } from "../../.claude/skills/level0/lib/schema-read.js";
import { cellsOf, rowsIn } from "../../.claude/skills/level0/lib/schema-table.js";
import { found, good, messages, NOTE, rules, SCHEMA, shown, swap } from "./schema-notes.js";

const withTable = (table = shown) =>
  good.replace("# What stands open", `${table}# What stands open`);
const lineOf = (text, line) => text.split(/\n/)[line - 1];

test("the schema reads as a map, a list, a flow list and a link", () => {
  assert.equal(SCHEMA.kind, "note");
  assert.deepEqual(SCHEMA.frontmatter.required, ["kind", "status"]);
  assert.deepEqual(SCHEMA.frontmatter.properties.status.enum, ["todo", "held", "done"]);
  assert.equal(SCHEMA.frontmatter.properties.kind["x-link"], true);
  assert.equal(SCHEMA.body.sections.length, 4);
  assert.equal(SCHEMA.body.sections[1].maxItems, 2);
  assert.equal(SCHEMA.body.sections[1].subsections.numbered, true);
  assert.equal(readYaml("kind: [[guidance]]\n").kind, "[[guidance]]");
});

test("a note in the shape its schema names breaks no rule", () => {
  assert.deepEqual(found(good), []);
});

// [[spec/design_output/schema#the-door-refuses-a-departure]]
test("every finding carries the shape this tree prints, at error", () => {
  const one = found(swap("status: todo\n", ""))[0];
  assert.deepEqual(Object.keys(one).sort(), [
    "column",
    "file",
    "line",
    "message",
    "rule",
    "severity",
  ]);
  assert.equal(one.file, NOTE);
  assert.equal(one.column, 1);
  assert.equal(one.severity, SEVERITY);
  assert.equal(
    SEVERITY,
    "error",
    "every note stands at its shape, so a departure turns check red",
  );
});

test("a note short of a required field is refused", () => {
  assert.ok(rules(swap("status: todo\n", "")).includes("Schema.status"));
  assert.match(messages(swap("status: todo\n", ""))[0], /names status/);
});

test("a note carrying a field the schema never names is refused", () => {
  const bad = swap("status: todo\n", "status: todo\nabout: a thing\n");
  assert.ok(rules(bad).includes("Schema.about"));
  assert.match(messages(bad).join(" "), /names no about/);
  assert.equal(found(bad)[0].line, 4, "it points at the line holding the field");
});

test("a note whose kind reads as another kind is refused", () => {
  const bad = swap("kind: [[note]]", "kind: [[other]]");
  assert.ok(rules(bad).includes("Schema.kind"));
  assert.match(messages(bad).join(" "), /a note note names note/);
});

test("a field off the enum is refused, and one on it passes", () => {
  const bad = swap("status: todo", "status: later");
  assert.ok(rules(bad).includes("Schema.status"));
  assert.match(messages(bad).join(" "), /todo, held, done/);
  assert.deepEqual(found(swap("status: todo", "status: done")), []);
});

test("a field taking a list, written as one line, is refused", () => {
  const bad = swap("status: todo\n", "status: todo\nscope: one line\n");
  assert.ok(rules(bad).includes("Schema.scope"));
  assert.match(messages(bad).join(" "), /takes array/);
  assert.deepEqual(found(swap("status: todo\n", 'status: todo\nscope: ["all"]\n')), []);
});

test("a field naming a link, written bare, is refused", () => {
  const bad = swap("status: todo\n", "status: todo\nexplains: spec/guidance/voice\n");
  assert.ok(rules(bad).includes("Schema.explains"));
  assert.match(messages(bad).join(" "), /names a link/);
  const linked = swap(
    "status: todo\n",
    "status: todo\nexplains: [[spec/guidance/voice]]\n",
  );
  assert.deepEqual(found(linked), []);
});

test("a note opening with no frontmatter is refused", () => {
  const bad = good.replace(/^---\n[\s\S]*?\n---\n\n/, "");
  assert.deepEqual(rules(bad), ["Schema.Frontmatter"]);
});

test("a note short of a required chapter is refused", () => {
  const bad = swap("# Scope\n\nWhat this note covers.\n\n", "");
  assert.ok(rules(bad).includes("Schema.Scope"));
  assert.match(messages(bad).join(" "), /carries a Scope chapter/);
});

test("a chapter the schema never names is refused", () => {
  const bad = `${good}\n# Stranger\n\nA chapter nobody names.\n`;
  assert.ok(rules(bad).includes("Schema.Stranger"));
  assert.match(messages(bad).join(" "), /names no Stranger chapter/);
});

test("a chapter out of the order the schema names is refused", () => {
  const bad = good
    .replace("# Scope\n\nWhat this note covers.\n\n", "")
    .replace(
      "# Actionables",
      "# Actionables\n\nOne line.\n\n# Scope\n\nWhat this covers.\n\n# Actionables",
    );
  assert.match(messages(bad).join(" "), /stands after/);
});

test("a closing chapter with another after it is refused", () => {
  const bad = `${good}\n# Scope\n\nA second scope.\n`;
  assert.match(messages(bad).join(" "), /closes this note, and Scope stands after it/);
});

test("a chapter holding no list, where the schema names one, is refused", () => {
  const bad = swap(
    "1. Do the first thing.\n2. Do the second thing.",
    "Two things, said as prose.",
  );
  assert.ok(rules(bad).includes("Schema.Actionables"));
  assert.match(messages(bad).join(" "), /holds a list/);
});

test("a list the schema numbers, written with bullets, is refused", () => {
  const bad = swap(
    "1. Do the first thing.\n2. Do the second thing.",
    "- Do the first thing.\n- Do the second thing.",
  );
  assert.ok(rules(bad).includes("Schema.Actionables"));
  assert.match(messages(bad).join(" "), /numbers every item/);
});

test("a list past the cap is refused, and it points at the item past it", () => {
  const bad = swap(
    "2. Do the second thing.",
    "2. Do the second thing.\n3. Do the third thing.",
  );
  const one = found(bad).find((said) => /A note holds 2 items\./.test(said.message));
  assert.ok(one, "the cap answers the count the schema names");
  assert.equal(bad.split(/\n/)[one.line - 1], "3. Do the third thing.");
});

test("a chapter under a numbered one, opening with no number, is refused", () => {
  const bad = swap("## 1. The first", "## The first");
  assert.ok(rules(bad).includes("Schema.Actionables"));
  assert.match(messages(bad).join(" "), /opens with the number/);
});

test("numbered chapters running back down are refused", () => {
  const bad = swap("## 1. The first", "## 4. The fourth");
  assert.match(messages(bad).join(" "), /the numbers run up/);
});

// [[spec/design_output/schema#a-chapter-holds-a-table]]
test("a chapter holding the table its schema names, one row per item, passes", () => {
  assert.deepEqual(found(withTable()), []);
  assert.deepEqual(rowsIn(shown.split("\n")).map((one) => one.line), [3, 5, 6]);
  assert.deepEqual(cellsOf("| 1 | do | do not |"), ["1", "do", "do not"]);
});

test("a chapter holding no table, where the schema names one, is refused", () => {
  const bad = withTable("# Examples\n\nTwo rows, said as prose.\n\n");
  assert.ok(rules(bad).includes("Schema.Examples"));
  assert.match(messages(bad).join(" "), /holds a table headed the rule, do, do not/);
});

test("a table opening with other heads is refused, at the head row", () => {
  const bad = withTable(shown.replace("| the rule | do | do not |", "| rule | yes | no |"));
  const one = found(bad).find((said) => /opens with the heads/.test(said.message));
  assert.ok(one, "the heads answer the schema");
  assert.equal(lineOf(bad, one.line), "| rule | yes | no |");
});

test("a table takes several rows for one item and none for another", () => {
  const twice = withTable(shown.replace("| 2 | do the second", "| 1 | do the second"));
  assert.deepEqual(found(twice), []);
  const one = withTable(shown.replace("| 2 | do the second thing | leave it |\n", ""));
  assert.deepEqual(found(one), []);
});

test("a row naming no item of the list is refused, and rows running back down too", () => {
  const off = withTable(shown.replace("| 2 | do the second", "| 3 | do the second"));
  const one = found(off).find((said) =>
    /opens with the number of an item of Actionables/.test(said.message),
  );
  assert.ok(one, "the row names its item");
  assert.match(one.message, /opens with 3\./);
  assert.equal(lineOf(off, one.line), "| 3 | do the second thing | leave it |");
  const bare = withTable(shown.replace("| 2 | do the second", "| | do the second"));
  assert.match(messages(bare).join(" "), /opens with nothing\./);
  const back = withTable(
    shown
      .replace("| 1 | do the first", "| 2 | do the first")
      .replace("| 2 | do the second", "| 1 | do the second"),
  );
  assert.match(messages(back).join(" "), /for item 1 stands after one for item 2/);
});

test("a comment and a fenced block count toward no bound", () => {
  const said = itemsIn([
    "<!-- 1. a description mint writes -->",
    "1. A real item.",
    "```",
    "2. A line inside a fence.",
    "```",
  ]);
  assert.deepEqual(
    said.map((one) => one.said),
    ["1. A real item."],
  );
});

test("a minted note passes the checker of the kind it is minted from", () => {
  const text = mintNote(SCHEMA);
  assert.deepEqual(checkNote(text, SCHEMA, NOTE), []);
  assert.match(text, /^---\nkind: \[\[note\]\]\nstatus: todo\n---/);
  assert.match(text, /<!-- one rule per item -->/);
  assert.match(text, /\n1\. /, "a list section carries the item the schema names");
});

test("a note names its kind through the link its frontmatter carries", () => {
  assert.equal(kindOf(good), "note");
  assert.equal(kindOf("# No frontmatter\n"), "");
  assert.equal(readNote(good).sections[0].header, "Scope");
});

// [[spec/design_input/the-editor-draws-the-ticket#a-ticket-picks-a-process]]
test("an empty process holds back the keys the fill writes, and a missing one holds back nothing", () => {
  const schema = readYaml(`kind: ticket
frontmatter:
  type: object
  required: [kind, state, steps]
  properties:
    kind:
      const: ticket
      x-link: true
    state:
      enum: [draft, open]
      x-filled-by: process
    steps:
      type: array
      x-filled-by: process
    process:
      x-link: true
body:
  headingLevel: 1
  sections:
    - header: Ask
      required: true
`);
  const fresh = "---\nkind: [[ticket]]\nprocess:\n---\n\n# Ask\n\nA thing.\n";
  const keys = (text) =>
    checkNote(text, schema, "spec/tickets/a.md")
      .map((one) => one.rule)
      .filter((rule) => /state|steps/.test(rule))
      .sort();
  assert.deepEqual(keys(fresh), []);
  assert.deepEqual(keys(fresh.replace("process:\n", "")), ["Schema.state", "Schema.steps"]);
});

// [[spec/design_input/the-editor-draws-the-ticket#a-ticket-picks-a-process]]
test("a process the note names holds back nothing, so a missing route stands named", () => {
  const schema = readYaml(`kind: ticket
frontmatter:
  type: object
  required: [kind, steps]
  properties:
    kind:
      const: ticket
      x-link: true
    steps:
      type: array
      x-filled-by: process
    process:
      x-link: true
`);
  const said = checkNote("---\nkind: [[ticket]]\nprocess: [[spec/processes/trivial]]\n---\n", schema, "a.md");
  assert.ok(said.some((one) => one.rule === "Schema.steps"));
});

// [[spec/tickets/the-quoted-pair-stays-paired]]
test("a list item quoted whole reads as text, and a quoted key with a quoted value stays a pair", () => {
  const read = readYaml(
    'checklist:\n  - "no sentence says: a count, a member"\n  - \'single: text\'\n  - "a": "b"\n',
  );
  assert.deepEqual(read.checklist, [
    "no sentence says: a count, a member",
    "single: text",
    { '"a"': "b" },
  ]);
  assert.equal(yaml.quotedWhole('"a \\" b: c"'), true);
  assert.equal(yaml.quotedWhole('"a": "b"'), false);
});

File: src/lsp/schema_test.go
package main

import (
	"quackitect/yaml"

	"strings"
	"testing"
)

const handoverSchema = `kind: handover
governs:
  - HANDOVER.md
frontmatter:
  type: object
  additionalProperties: false
  required:
    - kind
    - status
  properties:
    kind:
      const: handover
      x-link: true
      description: the schema this note is minted from
    status:
      enum: [todo, held, done]
      description: where the work stands
    depends_on:
      type: [array, string]
      description: the branches this one waits for
body:
  headingLevel: 1
  order: strict
  extraSections: false
  sections:
    - header: Where it stands
      required: true
      description: what the last session finished
    - header: What waits
      required: true
      list: true
      ordered: true
      maxItems: 2
      description: what the next reader does
`

func schemaHere(t *testing.T) *yaml.Doc {
	t.Helper()
	said := yaml.AsDoc(yaml.Read(handoverSchema))
	if said == nil {
		t.Fatal("the schema reads as nothing")
	}
	return said
}

func TestANoteOpensWithFrontmatter(t *testing.T) {
	found := checkNote("# Where it stands\n", schemaHere(t), "HANDOVER.md")
	one := onlyOne(t, found, "Schema.Frontmatter")
	if !strings.Contains(one.Message, "opens with frontmatter") {
		t.Errorf("the message reads %q", one.Message)
	}
}

func TestARequiredFieldStandsNamed(t *testing.T) {
	found := checkNote("---\nkind: [[handover]]\n---\n\n# Where it stands\n\n# What waits\n\n1. Go.\n",
		schemaHere(t), "HANDOVER.md")
	if names(found, "Schema.status") != 1 {
		t.Fatalf("a missing status answers %v", rules(found))
	}
}

func TestAFieldOutsideTheEnumRefuses(t *testing.T) {
	found := checkNote("---\nkind: [[handover]]\nstatus: maybe\n---\n\n# Where it stands\n\n# What waits\n\n1. Go.\n",
		schemaHere(t), "HANDOVER.md")
	one := onlyOne(t, found, "Schema.status")
	if !strings.Contains(one.Message, "todo, held, done") {
		t.Errorf("the message reads %q", one.Message)
	}
	if one.Line != 3 {
		t.Errorf("the finding names line %d", one.Line)
	}
}

func TestAFieldTheSchemaNamesNotRefuses(t *testing.T) {
	found := checkNote("---\nkind: [[handover]]\nstatus: todo\nstranger: yes\n---\n\n# Where it stands\n\n# What waits\n\n1. Go.\n",
		schemaHere(t), "HANDOVER.md")
	one := onlyOne(t, found, "Schema.stranger")
	if !strings.Contains(one.Message, "names no stranger") {
		t.Errorf("the message reads %q", one.Message)
	}
}

const ticketSchema = `kind: ticket
frontmatter:
  type: object
  properties:
    kind:
      const: ticket
      x-link: true
    steps:
      type: array
body:
  headingLevel: 1
  order: strict
  extraSections: false
  sections:
    - header: Ask
      required: true
    - x-one-per: steps
    - header: Discussion
      required: true
      position: last
`

// [[spec/design_output/schema#keywords-that-name-a-step]]
func TestAStepNamesItsOwnChapter(t *testing.T) {
	schema := yaml.AsDoc(yaml.Read(ticketSchema))
	note := "---\nkind: [[ticket]]\nsteps:\n  - name: do\n    does: makes it\n---\n\n# Ask\n\nA thing.\n\n# do\n\nDone.\n\n# Discussion\n"
	if found := checkNote(note, schema, "spec/tickets/a.md"); len(found) != 0 {
		t.Fatalf("a chapter a step names answers %v", rules(found))
	}

	missing := strings.Replace(note, "# do\n\nDone.\n\n", "", 1)
	if names(checkNote(missing, schema, "spec/tickets/a.md"), "Schema.do") != 1 {
		t.Fatal("a step with no chapter answers nothing")
	}
}

func TestALinkFieldNamesALink(t *testing.T) {
	found := checkNote("---\nkind: handover\nstatus: todo\n---\n\n# Where it stands\n\n# What waits\n\n1. Go.\n",
		schemaHere(t), "HANDOVER.md")
	one := onlyOne(t, found, "Schema.kind")
	if !strings.Contains(one.Message, "names a link") {
		t.Errorf("the message reads %q", one.Message)
	}
}

func TestATypeHoldsTheField(t *testing.T) {
	found := checkNote("---\nkind: [[handover]]\nstatus: todo\ndepends_on: 3\n---\n\n# Where it stands\n\n# What waits\n\n1. Go.\n",
		schemaHere(t), "HANDOVER.md")
	one := onlyOne(t, found, "Schema.DependsOn")
	if !strings.Contains(one.Message, "array or string") {
		t.Errorf("the message reads %q", one.Message)
	}
}

func TestAMissingChapterRefuses(t *testing.T) {
	found := checkNote("---\nkind: [[handover]]\nstatus: todo\n---\n\n# Where it stands\n",
		schemaHere(t), "HANDOVER.md")
	one := onlyOne(t, found, "Schema.WhatWaits")
	if !strings.Contains(one.Message, "carries a What waits chapter") {
		t.Errorf("the message reads %q", one.Message)
	}
}

func TestAChapterTheSchemaNamesNotRefuses(t *testing.T) {
	found := checkNote("---\nkind: [[handover]]\nstatus: todo\n---\n\n# Where it stands\n\n# What waits\n\n1. Go.\n\n# Stranger\n",
		schemaHere(t), "HANDOVER.md")
	one := onlyOne(t, found, "Schema.Stranger")
	if one.Line != 12 {
		t.Errorf("the finding names line %d", one.Line)
	}
}

func TestChaptersRunInOrder(t *testing.T) {
	found := checkNote("---\nkind: [[handover]]\nstatus: todo\n---\n\n# What waits\n\n1. Go.\n\n# Where it stands\n",
		schemaHere(t), "HANDOVER.md")
	one := onlyOne(t, found, "Schema.WhereItStands")
	if !strings.Contains(one.Message, "puts it first") {
		t.Errorf("the message reads %q", one.Message)
	}
}

func TestAListChapterHoldsAList(t *testing.T) {
	found := checkNote("---\nkind: [[handover]]\nstatus: todo\n---\n\n# Where it stands\n\n# What waits\n\nProse alone.\n",
		schemaHere(t), "HANDOVER.md")
	one := onlyOne(t, found, "Schema.WhatWaits")
	if !strings.Contains(one.Message, "holds a list") {
		t.Errorf("the message reads %q", one.Message)
	}
}

func TestAnOrderedChapterNumbersEveryItem(t *testing.T) {
	found := checkNote("---\nkind: [[handover]]\nstatus: todo\n---\n\n# Where it stands\n\n# What waits\n\n- Go.\n",
		schemaHere(t), "HANDOVER.md")
	one := onlyOne(t, found, "Schema.WhatWaits")
	if !strings.Contains(one.Message, "numbers every item") {
		t.Errorf("the message reads %q", one.Message)
	}
}

func TestAChapterHoldsItsCap(t *testing.T) {
	found := checkNote("---\nkind: [[handover]]\nstatus: todo\n---\n\n# Where it stands\n\n# What waits\n\n1. One.\n2. Two.\n3. Three.\n",
		schemaHere(t), "HANDOVER.md")
	one := onlyOne(t, found, "Schema.WhatWaits")
	if !strings.Contains(one.Message, "holds 2 items") {
		t.Errorf("the message reads %q", one.Message)
	}
}

func TestAFenceHidesAHeadingAndAnItem(t *testing.T) {
	said := "---\nkind: [[handover]]\nstatus: todo\n---\n\n# Where it stands\n\n```\n# Stranger\n- not an item\n```\n\n# What waits\n\n1. Go.\n"
	if found := checkNote(said, schemaHere(t), "HANDOVER.md"); len(found) != 0 {
		t.Fatalf("a fenced heading answers %v", rules(found))
	}
}

func TestAPlaceholderStandsAtWarning(t *testing.T) {
	said := "---\nkind: [[handover]]\nstatus: todo\n---\n\n# Where it stands\n\n<!-- what the last session finished -->\n\n# What waits\n\n1. Go.\n"
	found := placeholderFaults(said, schemaHere(t), "HANDOVER.md")
	one := onlyOne(t, found, "Schema.Placeholder")
	if one.Severity != SeverityWarning {
		t.Errorf("a placeholder stands at %q", one.Severity)
	}
	if one.Line != 8 {
		t.Errorf("the finding names line %d", one.Line)
	}
}

func TestAStrangerKindMeetsTheGovernor(t *testing.T) {
	tree := fixture(t, map[string]string{
		"spec/schemas/handover.schema.yaml": handoverSchema,
		"HANDOVER.md":                       "---\nkind: [[rationale]]\n---\n\n# Why\n",
	})
	schemas := schemasIn(tree)
	found := noteFaults(tree, schemas, "HANDOVER.md", tree.Read("HANDOVER.md"))
	one := onlyOne(t, found, "Schema.Kind")
	if !strings.Contains(one.Message, "reads as a rationale") {
		t.Errorf("the message reads %q", one.Message)
	}
}

func TestAKindWithNoSchemaRefuses(t *testing.T) {
	tree := fixture(t, map[string]string{
		"spec/schemas/handover.schema.yaml": handoverSchema,
		"spec/one.md":                       "---\nkind: [[stranger]]\n---\n\n# Why\n",
	})
	found := noteFaults(tree, schemasIn(tree), "spec/one.md", tree.Read("spec/one.md"))
	one := onlyOne(t, found, "Schema.Kind")
	if !strings.Contains(one.Message, "names no schema") {
		t.Errorf("the message reads %q", one.Message)
	}
}

func TestTheSweepSkipsADraft(t *testing.T) {
	tree := fixture(t, map[string]string{
		"spec/schemas/handover.schema.yaml": handoverSchema,
		"_draft.md":                         "---\nkind: [[handover]]\n---\n\n# Nothing\n",
	})
	if found := schemaFaults(tree); len(found) != 0 {
		t.Fatalf("a parked draft answers %v", found)
	}
}

// [[spec/design_output/schema#a-schema-names-its-chapters]]
func TestASchemaNamingChaptersReadsAsANoteSchema(t *testing.T) {
	if !isNoteSchema(schemaHere(t)) {
		t.Error("a schema with a kind and sections names a note schema")
	}
	if isNoteSchema(yaml.AsDoc(yaml.Read("kind: handover\n"))) {
		t.Error("a schema naming no chapter is no note schema")
	}
	if isNoteSchema(nil) {
		t.Error("no schema is no note schema")
	}
}

// A required key the fill writes waits while the note writes an empty process. [[spec/design_input/the-editor-draws-the-ticket#a-ticket-picks-a-process]]
const fillSchema = `kind: ticket
frontmatter:
  type: object
  required: [kind, state, steps]
  properties:
    kind:
      const: ticket
      x-link: true
    state:
      enum: [draft, open]
      x-filled-by: process
    steps:
      type: array
      x-filled-by: process
    process:
      x-link: true
body:
  headingLevel: 1
  sections:
    - header: Ask
      required: true
`

func TestAnEmptyProcessHoldsBackTheKeysTheFillWrites(t *testing.T) {
	schema := yaml.AsDoc(yaml.Read(fillSchema))
	fresh := "---\nkind: [[ticket]]\nprocess:\n---\n\n# Ask\n\nA thing.\n"
	found := checkNote(fresh, schema, "spec/tickets/a.md")
	if names(found, "Schema.steps")+names(found, "Schema.state") != 0 {
		t.Fatalf("a fresh ticket answers %v", rules(found))
	}
	bare := strings.Replace(fresh, "process:\n", "", 1)
	found = checkNote(bare, schema, "spec/tickets/a.md")
	if names(found, "Schema.steps") != 1 || names(found, "Schema.state") != 1 {
		t.Fatalf("a ticket naming no process answers %v", rules(found))
	}
}

// A governed folder holds its own kind alone, so a page there draws a warning. [[spec/tickets/each-folder-holds-its-kind]]
func TestAPagePastTheFolderKindWarns(t *testing.T) {
	tree := fixture(t, map[string]string{
		"spec/schemas/handover.schema.yaml": strings.Replace(handoverSchema, "  - HANDOVER.md", "  - spec/funnel/**", 1),
		"spec/funnel/page.html":             "<html></html>\n",
	})
	one := onlyOne(t, schemaFaults(tree), "Schema.Folder")
	if one.File != "spec/funnel/page.html" || one.Severity != "warning" {
		t.Errorf("the finding reads %+v", one)
	}
	if !strings.Contains(one.Message, "handover") {
		t.Errorf("the message reads %q", one.Message)
	}
}
